tests/graphite_e2e_test.py

```diff
@@ -0,0 +1,231 @@
+            "--publish",
+            "80",
+            "--publish",
+            "2003",
+            "graphiteapp/graphite-statsd",
+        ]
+        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
+        if proc.returncode != 0:
+            pytest.fail(
+                "Docker command returned non-zero exit code.\n\n"
+                f"Command: {cmd!r}\n"
+                f"Exit code: {proc.returncode}\n\n"
+                f"Stdout:\n{proc.stdout}\n\n"
+                f"Stderr:\n{proc.stderr}\n"
+            )
+        container_id = proc.stdout.strip()
+
+        # Determine the randomly assigned host port for 80/tcp (HTTP)
+        inspect_cmd = [
+            "docker",
+            "inspect",
+            "-f",
+            '{{ (index (index .NetworkSettings.Ports "80/tcp") 0).HostPort }}',
+            container_id,
+        ]
+        inspect_proc = subprocess.run(inspect_cmd, capture_output=True, text=True, timeout=60)
+        if inspect_proc.returncode != 0:
+            pytest.fail(
+                "Docker inspect returned non-zero exit code.\n\n"
+                f"Command: {inspect_cmd!r}\n"
+                f"Exit code: {inspect_proc.returncode}\n\n"
+                f"Stdout:\n{inspect_proc.stdout}\n\n"
+                f"Stderr:\n{inspect_proc.stderr}\n"
+            )
+        http_port = inspect_proc.stdout.strip()
+
+        # Determine the randomly assigned host port for 2003/tcp (Carbon)
+        inspect_cmd = [
+            "docker",
+            "inspect",
+            "-f",
+            '{{ (index (index .NetworkSettings.Ports "2003/tcp") 0).HostPort }}',
+            container_id,
+        ]
+        inspect_proc = subprocess.run(inspect_cmd, capture_output=True, text=True, timeout=60)
+        if inspect_proc.returncode != 0:
+            pytest.fail(
+                "Docker inspect returned non-zero exit code.\n\n"
+                f"Command: {inspect_cmd!r}\n"
+                f"Exit code: {inspect_proc.returncode}\n\n"
+                f"Stdout:\n{inspect_proc.stdout}\n\n"
+                f"Stderr:\n{inspect_proc.stderr}\n"
+            )
+        carbon_port = int(inspect_proc.stdout.strip())
+
+        # Wait until Graphite HTTP responds
+        deadline = time.time() + 60
+        ready = False
+        while time.time() < deadline:
+            try:
+                with socket.create_connection(("localhost", int(http_port)), timeout=1):
+                    ready = True
+                    break
+            except OSError:
+                time.sleep(1)
+
+        if not ready:
+            pytest.fail("Graphite HTTP port did not become ready within timeout.")
+
+        # Wait a bit more for Graphite to fully initialize
+        time.sleep(5)
+
+        # Seed data into Graphite using the same pattern as datagen.sh
+        _seed_graphite_data(carbon_port)
+
+        # Wait for data to be written and available
+        time.sleep(5)
+
+        yield http_port
```

Comment:
The postgresql version of this yields also the container ID. It bothers me that these are essentially the same function yet they are inconsistent with their return/yield values. Second, why isn't the carbon port returned?